# Importing Packages

In [438]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sn

# Load the data

In [439]:
# Loaded the data using pandas library 

loan_data = pd.read_csv(r'C:\Users\sabar\Python July 2023\Machine Learning\Projects\Money Lending Analysis\loan.csv', encoding = "ISO-8859-1")

C:\Users\sabar\AppData\Local\Temp\ipykernel_1072\2463682726.py:3: DtypeWarning: Columns (47) have mixed types. Specify dtype option on import or set low_memory=False.
  loan_data = pd.read_csv(r'C:\Users\sabar\Python July 2023\Machine Learning\Projects\Money Lending Analysis\loan.csv', encoding = "ISO-8859-1")


In [440]:
# Showcasing last first five rows and first and last columns

loan_data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000         4,975.00   36 months   
1  1077430    1314167       2500         2500         2,500.00   60 months   
2  1077175    1313524       2400         2400         2,400.00   36 months   
3  1076863    1277178      10000        10000        10,000.00   36 months   
4  1075358    1311748       3000         3000         3,000.00   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT   24,000.00            Verified  Dec-11   Fully Paid   
1           RENT   30,000.00     Source Verified  Dec-11  Charged Off   
2           RENT   12,252.00        Not Verified  Dec-11   Fully Paid   
3           RENT   49,200.00     Source Verified  Dec-11   Fully Paid   
4           RENT   80,000.00     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state   dti  delinq_2yrs  \
0              Computer    860xx         AZ 27.65            0   
1                  bike    309xx         GA  1.00            0   
2  real estate business    606xx         IL  8.72            0   
3              personel    917xx         CA 20.00            0   
4              Personal    972xx         OR 17.94            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85               1                     NaN   
1           Apr-99               5                     NaN   
2           Nov-01               2                     NaN   
3           Feb-96               1                   35.00   
4           Jan-96               0                   38.00   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         3        0      13648     83.70%          9   
1                     NaN         3        0       1687      9.40%          4   
2                     NaN         2        0       2956     98.50%         10   
3                     NaN        10        0       5598        21%         37   
4                     NaN        15        0      27783     53.90%         38   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   f       0.00           0.00     5,863.16         5,833.84   
1                   f       0.00           0.00     1,008.71         1,008.71   
2                   f       0.00           0.00     3,005.67         3,005.67   
3                   f       0.00           0.00    12,231.89        12,231.89   
4                   f     524.06         524.06     3,513.33         3,513.33   

   total_rec_prncp  total_rec_int  tota

In [441]:
# Showcasing 100 rows and columns

pd.set_option('display.max_rows', 100, 'display.max_columns',100)
loan_data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000         4,975.00   36 months   
1  1077430    1314167       2500         2500         2,500.00   60 months   
2  1077175    1313524       2400         2400         2,400.00   36 months   
3  1076863    1277178      10000        10000        10,000.00   36 months   
4  1075358    1311748       3000         3000         3,000.00   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT   24,000.00            Verified  Dec-11   Fully Paid   
1           RENT   30,000.00     Source Verified  Dec-11  Charged Off   
2           RENT   12,252.00        Not Verified  Dec-11   Fully Paid   
3           RENT   49,200.00     Source Verified  Dec-11   Fully Paid   
4           RENT   80,000.00     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state   dti  delinq_2yrs  \
0              Computer    860xx         AZ 27.65            0   
1                  bike    309xx         GA  1.00            0   
2  real estate business    606xx         IL  8.72            0   
3              personel    917xx         CA 20.00            0   
4              Personal    972xx         OR 17.94            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85               1                     NaN   
1           Apr-99               5                     NaN   
2           Nov-01               2                     NaN   
3           Feb-96               1                   35.00   
4           Jan-96               0                   38.00   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         3        0      13648     83.70%          9   
1                     NaN         3        0       1687      9.40%          4   
2                     NaN         2        0       2956     98.50%         10   
3                     NaN        10        0       5598        21%         37   
4                     NaN        15        0      27783     53.90%         38   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   f       0.00           0.00     5,863.16         5,833.84   
1                   f       0.00           0.00     1,008.71         1,008.71   
2                   f       0.00           0.00     3,005.67         3,005.67   
3                   f       0.00           0.00    12,231.89        12,231.89   
4                   f     524.06         524.06     3,513.33         3,513.33   

   total_rec_prncp  total_rec_int  tota

In [442]:
# Showcasing total number of rows and columns

loan_data.shape

(39717, 111)

In [443]:
# describing the dataset

loan_data.describe()

id    member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
count    39,717.00    39,717.00  39,717.00    39,717.00        39,717.00   
mean    683,131.91   850,463.56  11,219.44    10,947.71        10,397.45   
std     210,694.13   265,678.31   7,456.67     7,187.24         7,128.45   
min      54,734.00    70,699.00     500.00       500.00             0.00   
25%     516,221.00   666,780.00   5,500.00     5,400.00         5,000.00   
50%     665,665.00   850,812.00  10,000.00     9,600.00         8,975.00   
75%     837,755.00 1,047,339.00  15,000.00    15,000.00        14,400.00   
max   1,077,501.00 1,314,167.00  35,000.00    35,000.00        35,000.00   

       installment   annual_inc       dti  delinq_2yrs  inq_last_6mths  \
count    39,717.00    39,717.00 39,717.00    39,717.00       39,717.00   
mean        324.56    68,968.93     13.32         0.15            0.87   
std         208.87    63,793.77      6.68         0.49            1.07   
min          15.69     4,000.00      0.00         0.00            0.00   
25%         167.02    40,404.00      8.17         0.00            0.00   
50%         280.22    59,000.00     13.40         0.00            1.00   
75%         430.78    82,300.00     18.60         0.00            1.00   
max       1,305.19 6,000,000.00     29.99        11.00            8.00   

       mths_since_last_delinq  mths_since_last_record  open_acc   pub_rec  \
count               14,035.00                2,786.00 39,717.00 39,717.00   
mean                    35.90                   69.70      9.29      0.06   
std                     22.02                   43.82      4.40      0.24   
min                      0.00                    0.00      2.00      0.00   
25%                     18.00                   22.00      6.00      0.00   
50%                     34.00                   90.00      9.00      0.00   
75%                     52.00                  104.00     12.00      0.00   
max                    120.00                  129.00     44.00      4.00   

       revol_bal  total_acc  out_prncp  out_prncp_inv  total_pymnt  \
count  39,717.00  39,717.00  39,717.00      39,717.00    39,717.00   
mean   13,382.53      22.09      51.23          50.99    12,153.60   
std    15,885.02      11.40     375.17         373.82     9,042.04   
min         0.00       2.00       0.00           0.00         0.00   
25%     3,703.00      13.00       0.00           0.00     5,576.93   
50%     8,850.00      20.00       0.00           0.00     9,899.64   
75%    17,058.00      29.00       0.00           0.00    16,534.43   
max   149,588.00      90.00   6,311.47       6,307.37    58,563.68   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
count        39,717.00        39,717.00      39,717.00           39,717.00   
mean         11,567.15         9,793.35       2,263.66                1.36   
std           8,942.67         7,065.52       2,608.11                7.29   
min               0.00             0.00           0.00                0.00   
25%           5,112.31         4,600.00         662.18                0.00   
50%           9,287.15         8,000.00       1,348.91                0.00   
75%          15,798.81        13,653.26       2,833.40                0.00   
max          58,563.68        35,000.02      23,563.68              180.20   

       recoveries  collection_recovery_fee  last_pymnt_amnt  \
count   39,717.00                39,717.00        39,717.00   
mean        95.22                    12.41         2,678.83   
std        688.74                   148.67         4,447.14   
min          0.00                     0.00             0.00   
25%          0.00                     0.00           218.68   
50%          0.00                     0.00           546.14   
75%          0.00                     0.00         3,293.16   
max     29,623.35                 7,002.19        36,115.20   

       collections_12_mths_ex_med  mths_since_last_major_derog  poli

In [444]:
# Changing the data format

pd.options.display.float_format = '{:,.2f}'.format

In [445]:
# decribing the data after changing the format

loan_data.describe()

id    member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
count    39,717.00    39,717.00  39,717.00    39,717.00        39,717.00   
mean    683,131.91   850,463.56  11,219.44    10,947.71        10,397.45   
std     210,694.13   265,678.31   7,456.67     7,187.24         7,128.45   
min      54,734.00    70,699.00     500.00       500.00             0.00   
25%     516,221.00   666,780.00   5,500.00     5,400.00         5,000.00   
50%     665,665.00   850,812.00  10,000.00     9,600.00         8,975.00   
75%     837,755.00 1,047,339.00  15,000.00    15,000.00        14,400.00   
max   1,077,501.00 1,314,167.00  35,000.00    35,000.00        35,000.00   

       installment   annual_inc       dti  delinq_2yrs  inq_last_6mths  \
count    39,717.00    39,717.00 39,717.00    39,717.00       39,717.00   
mean        324.56    68,968.93     13.32         0.15            0.87   
std         208.87    63,793.77      6.68         0.49            1.07   
min          15.69     4,000.00      0.00         0.00            0.00   
25%         167.02    40,404.00      8.17         0.00            0.00   
50%         280.22    59,000.00     13.40         0.00            1.00   
75%         430.78    82,300.00     18.60         0.00            1.00   
max       1,305.19 6,000,000.00     29.99        11.00            8.00   

       mths_since_last_delinq  mths_since_last_record  open_acc   pub_rec  \
count               14,035.00                2,786.00 39,717.00 39,717.00   
mean                    35.90                   69.70      9.29      0.06   
std                     22.02                   43.82      4.40      0.24   
min                      0.00                    0.00      2.00      0.00   
25%                     18.00                   22.00      6.00      0.00   
50%                     34.00                   90.00      9.00      0.00   
75%                     52.00                  104.00     12.00      0.00   
max                    120.00                  129.00     44.00      4.00   

       revol_bal  total_acc  out_prncp  out_prncp_inv  total_pymnt  \
count  39,717.00  39,717.00  39,717.00      39,717.00    39,717.00   
mean   13,382.53      22.09      51.23          50.99    12,153.60   
std    15,885.02      11.40     375.17         373.82     9,042.04   
min         0.00       2.00       0.00           0.00         0.00   
25%     3,703.00      13.00       0.00           0.00     5,576.93   
50%     8,850.00      20.00       0.00           0.00     9,899.64   
75%    17,058.00      29.00       0.00           0.00    16,534.43   
max   149,588.00      90.00   6,311.47       6,307.37    58,563.68   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
count        39,717.00        39,717.00      39,717.00           39,717.00   
mean         11,567.15         9,793.35       2,263.66                1.36   
std           8,942.67         7,065.52       2,608.11                7.29   
min               0.00             0.00           0.00                0.00   
25%           5,112.31         4,600.00         662.18                0.00   
50%           9,287.15         8,000.00       1,348.91                0.00   
75%          15,798.81        13,653.26       2,833.40                0.00   
max          58,563.68        35,000.02      23,563.68              180.20   

       recoveries  collection_recovery_fee  last_pymnt_amnt  \
count   39,717.00                39,717.00        39,717.00   
mean        95.22                    12.41         2,678.83   
std        688.74                   148.67         4,447.14   
min          0.00                     0.00             0.00   
25%          0.00                     0.00           218.68   
50%          0.00                     0.00           546.14   
75%          0.00                     0.00         3,293.16   
max     29,623.35                 7,002.19        36,115.20   

       collections_12_mths_ex_med  mths_since_last_major_derog  poli

In [446]:
# information about the loand_data's index, columns, data types, null values, and memory usage.

loan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(74), int64(13), object(24)
memory usage: 33.6+ MB


# Data Cleaning!!

In [447]:
loan_data['loan_amnt'].describe()

count   39,717.00
mean    11,219.44
std      7,456.67
min        500.00
25%      5,500.00
50%     10,000.00
75%     15,000.00
max     35,000.00
Name: loan_amnt, dtype: float64

In [448]:
loan_data['loan_amnt'].isnull().sum()

0

In [449]:
loan_data['id'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
39712    False
39713    False
39714    False
39715    False
39716    False
Name: id, Length: 39717, dtype: bool

In [450]:
m = loan_data.isnull().sum()
m.values.tolist()

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 2459,
 1075,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 12940,
 0,
 11,
 0,
 0,
 0,
 0,
 0,
 0,
 25682,
 36931,
 0,
 0,
 0,
 50,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 71,
 0,
 38577,
 2,
 56,
 39717,
 0,
 0,
 39717,
 39717,
 39717,
 0,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 56,
 0,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 39717,
 697,
 39,
 39717,
 39717,
 39717,
 39717]

In [451]:
m1 = loan_data.index
m1

RangeIndex(start=0, stop=39717, step=1)

In [452]:
m2 = len(loan_data.index)
m2

39717

In [453]:
# Missing values 

# to find the missing data - (total number of null in a column / total number of columns)

missing_values = round(100*(loan_data.isnull().sum()/len(loan_data.index)), 2)
print(missing_values)

id                             0.00
member_id                      0.00
loan_amnt                      0.00
funded_amnt                    0.00
funded_amnt_inv                0.00
                              ...  
tax_liens                      0.10
tot_hi_cred_lim              100.00
total_bal_ex_mort            100.00
total_bc_limit               100.00
total_il_high_credit_limit   100.00
Length: 111, dtype: float64


In [454]:
missing_values.values.tolist()

[0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 6.19,
 2.71,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 32.58,
 0.0,
 0.03,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 64.66,
 92.99,
 0.0,
 0.0,
 0.0,
 0.13,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.18,
 0.0,
 97.13,
 0.01,
 0.14,
 100.0,
 0.0,
 0.0,
 100.0,
 100.0,
 100.0,
 0.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 0.14,
 0.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 100.0,
 1.75,
 0.1,
 100.0,
 100.0,
 100.0,
 100.0]

In [455]:
# Checking whether the data set has more than 90% is null

missing_values.loc[missing_values > 90].count()

56

In [456]:
column_with_missing_values = list(missing_values.loc[missing_values > 90].index)
len(column_with_missing_values)

56

In [457]:
# list of columns which as more than and equal to 90% null data

list(column_with_missing_values)

['mths_since_last_record',
 'next_pymnt_d',
 'mths_since_last_major_derog',
 'annual_inc_joint',
 'dti_joint',
 'verification_status_joint',
 'tot_coll_amt',
 'tot_cur_bal',
 'open_acc_6m',
 'open_il_6m',
 'open_il_12m',
 'open_il_24m',
 'mths_since_rcnt_il',
 'total_bal_il',
 'il_util',
 'open_rv_12m',
 'open_rv_24m',
 'max_bal_bc',
 'all_util',
 'total_rev_hi_lim',
 'inq_fi',
 'total_cu_tl',
 'inq_last_12m',
 'acc_open_past_24mths',
 'avg_cur_bal',
 'bc_open_to_buy',
 'bc_util',
 'mo_sin_old_il_acct',
 'mo_sin_old_rev_tl_op',
 'mo_sin_rcnt_rev_tl_op',
 'mo_sin_rcnt_tl',
 'mort_acc',
 'mths_since_recent_bc',
 'mths_since_recent_bc_dlq',
 'mths_since_recent_inq',
 'mths_since_recent_revol_delinq',
 'num_accts_ever_120_pd',
 'num_actv_bc_tl',
 'num_actv_rev_tl',
 'num_bc_sats',
 'num_bc_tl',
 'num_il_tl',
 'num_op_rev_tl',
 'num_rev_accts',
 'num_rev_tl_bal_gt_0',
 'num_sats',
 'num_tl_120dpd_2m',
 'num_tl_30dpd',
 'num_tl_90g_dpd_24m',
 'num_tl_op_past_12m',
 'pct_tl_nvr_dlq',
 'percen

In [458]:
# Gonna drop the data which has more than 90% null data

loan_data = loan_data.drop(column_with_missing_values, axis = 1)

In [459]:
loan_data.shape

(39717, 55)

## As we dropped the null columns (56), Now we have 55 columns previously it was 111 columns

In [460]:
missing_values = round(100*(loan_data.isnull().sum()/len(loan_data.index)), 2)
column_with_missing_values = missing_values.loc[missing_values > 50].index
column_with_missing_values

Index(['mths_since_last_delinq'], dtype='object')

In [461]:
list(column_with_missing_values)

['mths_since_last_delinq']

In [462]:
column_with_missing_values.values.tolist()

['mths_since_last_delinq']

In [463]:
# object is string

loan_data.loc[:, loan_data.dtypes == np.object].columns

C:\Users\sabar\AppData\Local\Temp\ipykernel_1072\3356870952.py:3: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  loan_data.loc[:, loan_data.dtypes == np.object].columns


Index(['term', 'int_rate', 'grade', 'sub_grade', 'emp_title', 'emp_length',
       'home_ownership', 'verification_status', 'issue_d', 'loan_status',
       'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code',
       'addr_state', 'earliest_cr_line', 'revol_util', 'initial_list_status',
       'last_pymnt_d', 'last_credit_pull_d', 'application_type'],
      dtype='object')

In [464]:
# Dropping two columns as those data are not important for analysis

# loan_data = loan_data.drop(['url','desc'], axis=1)

In [465]:
loan_data.shape

(39717, 55)

In [466]:
loan_data.dtypes

id                              int64
member_id                       int64
loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                        object
loan_status                    object
pymnt_plan                     object
url                            object
desc                           object
purpose                        object
title                          object
zip_code                       object
addr_state                     object
dti                           float64
delinq_2yrs                     int64
earliest_cr_

In [467]:
# getting unique values 

UniqueDataColumns = []
for rows in loan_data.columns.tolist():
    UniqueDataColumns.append({'column': rows, 'count': len(loan_data[rows].unique())})
UniqueData = pd.DataFrame(UniqueDataColumns).sort_values('count')
UniqueData

column  count
52                 delinq_amnt      1
50              acc_now_delinq      1
49            application_type      1
48                 policy_code      1
34         initial_list_status      1
17                  pymnt_plan      1
54                   tax_liens      2
51    chargeoff_within_12_mths      2
5                         term      2
47  collections_12_mths_ex_med      2
14         verification_status      3
16                 loan_status      3
53        pub_rec_bankruptcies      4
30                     pub_rec      5
12              home_ownership      5
8                        grade      7
27              inq_last_6mths      9
25                 delinq_2yrs     11
11                  emp_length     12
20                     purpose     14
9                    sub_grade     35
29                    open_acc     40
23                  addr_state     50
15                     issue_d     55
33                   total_acc     82
28      mths_since_last_delinq     96
44                last_pymnt_d    102
46          last_credit_pull_d    107
6                     int_rate    371
26            earliest_cr_line    526
22                    zip_code    823
2                    loan_amnt    885
3                  funded_amnt   1041
32                  revol_util   1090
35                   out_prncp   1137
36               out_prncp_inv   1138
41          total_rec_late_fee   1356
43     collection_recovery_fee   2616
24                         dti   2868
42                  recoveries   4040
13                  annual_inc   5318
39             total_rec_prncp   7976
4              funded_amnt_inv   8205
7                  installment  15383
21                       title  19616
31                   revol_bal  21711
19                        desc  26528
10                   emp_title  28821
45             last_pymnt_amnt  34930
40               total_rec_int  35148
38             total_pymnt_inv  37518
37                 total_pymnt  37850
18                         url  39717
1                    member_id  39717
0                           id  39717

In [468]:
len(UniqueData)

55

In [469]:
loan_data

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0      1077501    1296599       5000         5000         4,975.00   
1      1077430    1314167       2500         2500         2,500.00   
2      1077175    1313524       2400         2400         2,400.00   
3      1076863    1277178      10000        10000        10,000.00   
4      1075358    1311748       3000         3000         3,000.00   
...        ...        ...        ...          ...              ...   
39712    92187      92174       2500         2500         1,075.00   
39713    90665      90607       8500         8500           875.00   
39714    90395      90390       5000         5000         1,325.00   
39715    90376      89243       5000         5000           650.00   
39716    87023      86999       7500         7500           800.00   

             term int_rate  installment grade sub_grade  \
0       36 months   10.65%       162.87     B        B2   
1       60 months   15.27%        59.83     C        C4   
2       36 months   15.96%        84.33     C        C5   
3       36 months   13.49%       339.31     C        C1   
4       60 months   12.69%        67.79     B        B5   
...           ...      ...          ...   ...       ...   
39712   36 months    8.07%        78.42     A        A4   
39713   36 months   10.28%       275.38     C        C1   
39714   36 months    8.07%       156.84     A        A4   
39715   36 months    7.43%       155.38     A        A2   
39716   36 months   13.75%       255.43     E        E2   

                        emp_title emp_length home_ownership  annual_inc  \
0                             NaN  10+ years           RENT   24,000.00   
1                           Ryder   < 1 year           RENT   30,000.00   
2                             NaN  10+ years           RENT   12,252.00   
3             AIR RESOURCES BOARD  10+ years           RENT   49,200.00   
4        University Medical Group     1 year           RENT   80,000.00   
...                           ...        ...            ...         ...   
39712             FiSite Research    4 years       MORTGAGE  110,000.00   
39713  Squarewave Solutions, Ltd.    3 years           RENT   18,000.00   
39714                         NaN   < 1 year       MORTGAGE  100,000.00   
39715                         NaN   < 1 year       MORTGAGE  200,000.00   
39716            Evergreen Center   < 1 year            OWN   22,000.00   

      verification_status issue_d  loan_status pymnt_plan  \
0                Verified  Dec-11   Fully Paid          n   
1         Source Verified  Dec-11  Charged Off          n   
2            Not Verified  Dec-11   Fully Paid          n   
3         Source Verified  Dec-11   Fully Paid          n   
4         Source Verified  Dec-11      Current          n   
...                   ...     ...          ...        ...   
39712        Not Verified  Jul-07   Fully Paid          n   
39713        Not Verified  Jul-07   Fully Paid          n   
39714        Not Verified  Jul-07   Fully Paid          n   
39715        Not Verified  Jul-07   Fully Paid          n   
39716        Not Verified  Jun-07   Fully Paid          n   

                                                     url  \
0      https://lendingclub.com/browse/loanDetail.acti...   
1      https://lendingclub.com/browse/loanDetail.acti...   
2      https://lendingclub.com/browse/loanDetail.acti...   
3      https://lendingclub.com/browse/loanDetail.acti...   
4      https://lendingclub.com/browse/loanDetail.acti...   
...                                                  ...   
39712  https://lendingclub.com/browse/loanDetail.acti...   
39713  https://lendingclub.com/browse/loanDetail.acti...   
39714  https://lendingclub.com/browse/loanDetail.acti...   
39715  https://lendingclub.com/browse/loanDetail.acti...   
39716  https://lendingclub.com/browse/loanDetail.acti...   

                                                    desc             purpose  \
0        Borrower added on 

In [470]:
loan_data['delinq_amnt'].unique()

array([0], dtype=int64)

In [471]:
loan_data['acc_now_delinq'].unique()

array([0], dtype=int64)

In [472]:
loan_data['application_type'].unique()

array(['INDIVIDUAL'], dtype=object)

In [473]:
loan_data['policy_code'].unique()

array([1], dtype=int64)

In [474]:
loan_data['initial_list_status'].unique()

array(['f'], dtype=object)

In [475]:
loan_data['pymnt_plan'].unique()

array(['n'], dtype=object)

In [476]:
loan_data['tax_liens'].unique()

array([ 0., nan])

In [477]:
loan_data['chargeoff_within_12_mths'].unique()

array([ 0., nan])

In [478]:
loan_data['term'].unique()

array([' 36 months', ' 60 months'], dtype=object)

In [479]:
loan_data['collections_12_mths_ex_med'].unique()

array([ 0., nan])

In [480]:
loan_data['loan_status'].unique()

array(['Fully Paid', 'Charged Off', 'Current'], dtype=object)

In [481]:
# Most of the values are 0, Null. so gonna drop now
# single unique data is not helpful for our analysis. so gonna drop those data as well

unwanted_columns = ['delinq_amnt','acc_now_delinq','application_type','policy_code','initial_list_status','pymnt_plan',
                    'tax_liens','chargeoff_within_12_mths']

In [482]:
loan_data.shape

(39717, 55)

In [483]:
loan_data = loan_data.drop(unwanted_columns, axis=1)
loan_data.shape

(39717, 47)

In [484]:
missing_values = round(100*(loan_data.isnull().sum()/len(loan_data.index)), 2)
missing_values

id                            0.00
member_id                     0.00
loan_amnt                     0.00
funded_amnt                   0.00
funded_amnt_inv               0.00
term                          0.00
int_rate                      0.00
installment                   0.00
grade                         0.00
sub_grade                     0.00
emp_title                     6.19
emp_length                    2.71
home_ownership                0.00
annual_inc                    0.00
verification_status           0.00
issue_d                       0.00
loan_status                   0.00
url                           0.00
desc                         32.58
purpose                       0.00
title                         0.03
zip_code                      0.00
addr_state                    0.00
dti                           0.00
delinq_2yrs                   0.00
earliest_cr_line              0.00
inq_last_6mths                0.00
mths_since_last_delinq       64.66
open_acc            

In [485]:
MV1 = missing_values.loc[missing_values > 50].index
MV1

Index(['mths_since_last_delinq'], dtype='object')

In [486]:
# Dropping this column, as it has more than 60% of Null data

loan_data = loan_data.drop(['mths_since_last_delinq'], axis = 1)
loan_data.shape

(39717, 46)

In [487]:
loan_data = loan_data.drop(['id', 'member_id'], axis = 1)
loan_data.shape

(39717, 44)

# Validate the data type!!

In [488]:
loan_data.head(10)

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       5000         5000         4,975.00   36 months   10.65%       162.87   
1       2500         2500         2,500.00   60 months   15.27%        59.83   
2       2400         2400         2,400.00   36 months   15.96%        84.33   
3      10000        10000        10,000.00   36 months   13.49%       339.31   
4       3000         3000         3,000.00   60 months   12.69%        67.79   
5       5000         5000         5,000.00   36 months    7.90%       156.46   
6       7000         7000         7,000.00   60 months   15.96%       170.08   
7       3000         3000         3,000.00   36 months   18.64%       109.43   
8       5600         5600         5,600.00   60 months   21.28%       152.39   
9       5375         5375         5,350.00   60 months   12.69%       121.45   

  grade sub_grade                  emp_title emp_length home_ownership  \
0     B        B2                        NaN  10+ years           RENT   
1     C        C4                      Ryder   < 1 year           RENT   
2     C        C5                        NaN  10+ years           RENT   
3     C        C1        AIR RESOURCES BOARD  10+ years           RENT   
4     B        B5   University Medical Group     1 year           RENT   
5     A        A4       Veolia Transportaton    3 years           RENT   
6     C        C5  Southern Star Photography    8 years           RENT   
7     E        E1            MKC Accounting     9 years           RENT   
8     F        F2                        NaN    4 years            OWN   
9     B        B5                  Starbucks   < 1 year           RENT   

   annual_inc verification_status issue_d  loan_status  \
0   24,000.00            Verified  Dec-11   Fully Paid   
1   30,000.00     Source Verified  Dec-11  Charged Off   
2   12,252.00        Not Verified  Dec-11   Fully Paid   
3   49,200.00     Source Verified  Dec-11   Fully Paid   
4   80,000.00     Source Verified  Dec-11      Current   
5   36,000.00     Source Verified  Dec-11   Fully Paid   
6   47,004.00        Not Verified  Dec-11   Fully Paid   
7   48,000.00     Source Verified  Dec-11   Fully Paid   
8   40,000.00     Source Verified  Dec-11  Charged Off   
9   15,000.00            Verified  Dec-11  Charged Off   

                                                 url  \
0  https://lendingclub.com/browse/loanDetail.acti...   
1  https://lendingclub.com/browse/loanDetail.acti...   
2  https://lendingclub.com/browse/loanDetail.acti...   
3  https://lendingclub.com/browse/loanDetail.acti...   
4  https://lendingclub.com/browse/loanDetail.acti...   
5  https://lendingclub.com/browse/loanDetail.acti...   
6  https://lendingclub.com/browse/loanDetail.acti...   
7  https://lendingclub.com/browse/loanDetail.acti...   
8  https://lendingclub.com/browse/loanDetail.acti...   
9  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc             purpose  \
0    Borrower added on 12/22/11 > I need to upgra...         credit_card   
1    Borrower added on 12/22/11 > I plan to use t...                 car   
2                                                NaN      small_business   
3    Borrower added on 12/21/11 > to pay for prop...               other   
4    Borrower added on 12/21/11 > I plan on combi...               other   
5                                                NaN             wedding   
6    Borrower added on 12/18/11 > I am planning o...  debt_consolidation   
7    Borrower added on 12/16/11 > Downpayment for...                 car   
8    Borrower added on 12/21/11 > I own a small h...      small_business   
9    Borrower added on 12/16/11 > I'm trying to b...               other   

                                   title zip_code addr_state   dti  \
0                               Computer    860xx         AZ 27.65   
1                                   bike    309xx         GA  1.00   
2   

In [489]:
loan_data.tail(10)

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
39707       5000         5000           525.00   36 months    9.33%   
39708       5000         5000           375.00   36 months    9.96%   
39709       5000         5000           675.00   36 months   11.22%   
39710       5000         5000           250.00   36 months    7.43%   
39711       5000         5000           700.00   36 months    8.70%   
39712       2500         2500         1,075.00   36 months    8.07%   
39713       8500         8500           875.00   36 months   10.28%   
39714       5000         5000         1,325.00   36 months    8.07%   
39715       5000         5000           650.00   36 months    7.43%   
39716       7500         7500           800.00   36 months   13.75%   

       installment grade sub_grade                   emp_title emp_length  \
39707       159.77     B        B3          Stark and Roth Inc    2 years   
39708       161.25     B        B5             Millenium Group    4 years   
39709       164.23     C        C4              Self-Employeed   < 1 year   
39710       155.38     A        A2           Rush Univ Med Grp     1 year   
39711       158.30     B        B1         A. F. Wolfers, Inc.    5 years   
39712        78.42     A        A4             FiSite Research    4 years   
39713       275.38     C        C1  Squarewave Solutions, Ltd.    3 years   
39714       156.84     A        A4                         NaN   < 1 year   
39715       155.38     A        A2                         NaN   < 1 year   
39716       255.43     E        E2            Evergreen Center   < 1 year   

      home_ownership  annual_inc verification_status issue_d loan_status  \
39707       MORTGAGE  180,000.00        Not Verified  Jul-07  Fully Paid   
39708       MORTGAGE   48,000.00        Not Verified  Jul-07  Fully Paid   
39709            OWN   80,000.00        Not Verified  Jul-07  Fully Paid   
39710            OWN   85,000.00        Not Verified  Jul-07  Fully Paid   
39711       MORTGAGE   75,000.00        Not Verified  Jul-07  Fully Paid   
39712       MORTGAGE  110,000.00        Not Verified  Jul-07  Fully Paid   
39713           RENT   18,000.00        Not Verified  Jul-07  Fully Paid   
39714       MORTGAGE  100,000.00        Not Verified  Jul-07  Fully Paid   
39715       MORTGAGE  200,000.00        Not Verified  Jul-07  Fully Paid   
39716            OWN   22,000.00        Not Verified  Jun-07  Fully Paid   

                                                     url  \
39707  https://lendingclub.com/browse/loanDetail.acti...   
39708  https://lendingclub.com/browse/loanDetail.acti...   
39709  https://lendingclub.com/browse/loanDetail.acti...   
39710  https://lendingclub.com/browse/loanDetail.acti...   
39711  https://lendingclub.com/browse/loanDetail.acti...   
39712  https://lendingclub.com/browse/loanDetail.acti...   
39713  https://lendingclub.com/browse/loanDetail.acti...   
39714  https://lendingclub.com/browse/loanDetail.acti...   
39715  https://lendingclub.com/browse/loanDetail.acti...   
39716  https://lendingclub.com/browse/loanDetail.acti...   

                                                    desc             purpose  \
39707          Need a loan to make some home improvments    home_improvement   
39708  I would like to pay off my high-interest credi...  debt_consolidation   
39709                                                NaN         credit_card   
39710                                                NaN         credit_card   
39711  I'd like to shift some credit card debt so it ...         credit_card   
39712  Our current gutter system on our home is old a...    home_improvement   
39713  The rate of interest and fees incurred by carr...         credit_card   
39714                                                NaN  debt_consolidation   
39715                                                NaN               other   
39716  I plan to consolidate over $7,000 of debt: a c...  debt_consolidation   

In [490]:
loan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 44 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   loan_amnt                   39717 non-null  int64  
 1   funded_amnt                 39717 non-null  int64  
 2   funded_amnt_inv             39717 non-null  float64
 3   term                        39717 non-null  object 
 4   int_rate                    39717 non-null  object 
 5   installment                 39717 non-null  float64
 6   grade                       39717 non-null  object 
 7   sub_grade                   39717 non-null  object 
 8   emp_title                   37258 non-null  object 
 9   emp_length                  38642 non-null  object 
 10  home_ownership              39717 non-null  object 
 11  annual_inc                  39717 non-null  float64
 12  verification_status         39717 non-null  object 
 13  issue_d                     397

# Data Preparation

In [491]:
loan_data['term_months'] = loan_data['term'].str.strip('months')
del loan_data['term']
loan_data.head()

loan_amnt  funded_amnt  funded_amnt_inv int_rate  installment grade  \
0       5000         5000         4,975.00   10.65%       162.87     B   
1       2500         2500         2,500.00   15.27%        59.83     C   
2       2400         2400         2,400.00   15.96%        84.33     C   
3      10000        10000        10,000.00   13.49%       339.31     C   
4       3000         3000         3,000.00   12.69%        67.79     B   

  sub_grade                 emp_title emp_length home_ownership  annual_inc  \
0        B2                       NaN  10+ years           RENT   24,000.00   
1        C4                     Ryder   < 1 year           RENT   30,000.00   
2        C5                       NaN  10+ years           RENT   12,252.00   
3        C1       AIR RESOURCES BOARD  10+ years           RENT   49,200.00   
4        B5  University Medical Group     1 year           RENT   80,000.00   

  verification_status issue_d  loan_status  \
0            Verified  Dec-11   Fully Paid   
1     Source Verified  Dec-11  Charged Off   
2        Not Verified  Dec-11   Fully Paid   
3     Source Verified  Dec-11   Fully Paid   
4     Source Verified  Dec-11      Current   

                                                 url  \
0  https://lendingclub.com/browse/loanDetail.acti...   
1  https://lendingclub.com/browse/loanDetail.acti...   
2  https://lendingclub.com/browse/loanDetail.acti...   
3  https://lendingclub.com/browse/loanDetail.acti...   
4  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state   dti  delinq_2yrs  \
0              Computer    860xx         AZ 27.65            0   
1                  bike    309xx         GA  1.00            0   
2  real estate business    606xx         IL  8.72            0   
3              personel    917xx         CA 20.00            0   
4              Personal    972xx         OR 17.94            0   

  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal revol_util  \
0           Jan-85               1         3        0      13648     83.70%   
1           Apr-99               5         3        0       1687      9.40%   
2           Nov-01               2         2        0       2956     98.50%   
3           Feb-96               1        10        0       5598        21%   
4           Jan-96               0        15        0      27783     53.90%   

   total_acc  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0          9       0.00           0.00     5,863.16         5,833.84   
1          4       0.00           0.00     1,008.71         1,008.71   
2         10       0.00           0.00     3,005.67         3,005.67   
3         37       0.00           0.00    12,231.89        12,231.89   
4         38     524.06         524.06     3,513.33         3,513.33   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0         5,000.00         863.16                0.00        0.00   
1           456.46         435.17                0.00      117.08   
2         2,400.00         605.67                0.00        0.00   
3        10,000.00       2,214.92               16.97        0.00   
4         2,475.94       1,037.39                0.00        0.00   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt last_credit_pull_d  \
0                     0.00       Jan-15           171.62             May-16   
1                     1.11       Apr-13           119.66             Sep-13   
2                     0.00       Jun-14           649.91             May-16   


In [492]:
loan_data.head()

loan_amnt  funded_amnt  funded_amnt_inv int_rate  installment grade  \
0       5000         5000         4,975.00   10.65%       162.87     B   
1       2500         2500         2,500.00   15.27%        59.83     C   
2       2400         2400         2,400.00   15.96%        84.33     C   
3      10000        10000        10,000.00   13.49%       339.31     C   
4       3000         3000         3,000.00   12.69%        67.79     B   

  sub_grade                 emp_title emp_length home_ownership  annual_inc  \
0        B2                       NaN  10+ years           RENT   24,000.00   
1        C4                     Ryder   < 1 year           RENT   30,000.00   
2        C5                       NaN  10+ years           RENT   12,252.00   
3        C1       AIR RESOURCES BOARD  10+ years           RENT   49,200.00   
4        B5  University Medical Group     1 year           RENT   80,000.00   

  verification_status issue_d  loan_status  \
0            Verified  Dec-11   Fully Paid   
1     Source Verified  Dec-11  Charged Off   
2        Not Verified  Dec-11   Fully Paid   
3     Source Verified  Dec-11   Fully Paid   
4     Source Verified  Dec-11      Current   

                                                 url  \
0  https://lendingclub.com/browse/loanDetail.acti...   
1  https://lendingclub.com/browse/loanDetail.acti...   
2  https://lendingclub.com/browse/loanDetail.acti...   
3  https://lendingclub.com/browse/loanDetail.acti...   
4  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state   dti  delinq_2yrs  \
0              Computer    860xx         AZ 27.65            0   
1                  bike    309xx         GA  1.00            0   
2  real estate business    606xx         IL  8.72            0   
3              personel    917xx         CA 20.00            0   
4              Personal    972xx         OR 17.94            0   

  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal revol_util  \
0           Jan-85               1         3        0      13648     83.70%   
1           Apr-99               5         3        0       1687      9.40%   
2           Nov-01               2         2        0       2956     98.50%   
3           Feb-96               1        10        0       5598        21%   
4           Jan-96               0        15        0      27783     53.90%   

   total_acc  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0          9       0.00           0.00     5,863.16         5,833.84   
1          4       0.00           0.00     1,008.71         1,008.71   
2         10       0.00           0.00     3,005.67         3,005.67   
3         37       0.00           0.00    12,231.89        12,231.89   
4         38     524.06         524.06     3,513.33         3,513.33   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0         5,000.00         863.16                0.00        0.00   
1           456.46         435.17                0.00      117.08   
2         2,400.00         605.67                0.00        0.00   
3        10,000.00       2,214.92               16.97        0.00   
4         2,475.94       1,037.39                0.00        0.00   

   collection_recovery_fee last_pymnt_d  last_pymnt_amnt last_credit_pull_d  \
0                     0.00       Jan-15           171.62             May-16   
1                     1.11       Apr-13           119.66             Sep-13   
2                     0.00       Jun-14           649.91             May-16   


In [493]:
loan_data.shape

(39717, 44)

In [494]:
loan_data = loan_data.drop(['url', 'desc'], axis = 1)
loan_data.shape

(39717, 42)

In [495]:
loan_data.head(1)

loan_amnt  funded_amnt  funded_amnt_inv int_rate  installment grade  \
0       5000         5000         4,975.00   10.65%       162.87     B   

  sub_grade emp_title emp_length home_ownership  annual_inc  \
0        B2       NaN  10+ years           RENT   24,000.00   

  verification_status issue_d loan_status      purpose     title zip_code  \
0            Verified  Dec-11  Fully Paid  credit_card  Computer    860xx   

  addr_state   dti  delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  \
0         AZ 27.65            0           Jan-85               1         3   

   pub_rec  revol_bal revol_util  total_acc  out_prncp  out_prncp_inv  \
0        0      13648     83.70%          9       0.00           0.00   

   total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
0     5,863.16         5,833.84         5,000.00         863.16   

   total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
0                0.00        0.00                     0.00       Jan-15   

   last_pymnt_amnt last_credit_pull_d  collections_12_mths_ex_med  \
0           171.62             May-16                        0.00   

   pub_rec_bankruptcies term_months  
0                  0.00         36

In [496]:
loan_data['zip_code']=loan_data['zip_code'].str.rstrip('xx')
loan_data.head(1)

loan_amnt  funded_amnt  funded_amnt_inv int_rate  installment grade  \
0       5000         5000         4,975.00   10.65%       162.87     B   

  sub_grade emp_title emp_length home_ownership  annual_inc  \
0        B2       NaN  10+ years           RENT   24,000.00   

  verification_status issue_d loan_status      purpose     title zip_code  \
0            Verified  Dec-11  Fully Paid  credit_card  Computer      860   

  addr_state   dti  delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  \
0         AZ 27.65            0           Jan-85               1         3   

   pub_rec  revol_bal revol_util  total_acc  out_prncp  out_prncp_inv  \
0        0      13648     83.70%          9       0.00           0.00   

   total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
0     5,863.16         5,833.84         5,000.00         863.16   

   total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
0                0.00        0.00                     0.00       Jan-15   

   last_pymnt_amnt last_credit_pull_d  collections_12_mths_ex_med  \
0           171.62             May-16                        0.00   

   pub_rec_bankruptcies term_months  
0                  0.00         36

In [497]:
loan_data['int_rate_percentage'] = loan_data['int_rate'].str.strip('%')
loan_data.head(1)

loan_amnt  funded_amnt  funded_amnt_inv int_rate  installment grade  \
0       5000         5000         4,975.00   10.65%       162.87     B   

  sub_grade emp_title emp_length home_ownership  annual_inc  \
0        B2       NaN  10+ years           RENT   24,000.00   

  verification_status issue_d loan_status      purpose     title zip_code  \
0            Verified  Dec-11  Fully Paid  credit_card  Computer      860   

  addr_state   dti  delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  \
0         AZ 27.65            0           Jan-85               1         3   

   pub_rec  revol_bal revol_util  total_acc  out_prncp  out_prncp_inv  \
0        0      13648     83.70%          9       0.00           0.00   

   total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
0     5,863.16         5,833.84         5,000.00         863.16   

   total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
0                0.00        0.00                     0.00       Jan-15   

   last_pymnt_amnt last_credit_pull_d  collections_12_mths_ex_med  \
0           171.62             May-16                        0.00   

   pub_rec_bankruptcies term_months int_rate_percentage  
0                  0.00         36                10.65

In [498]:
loan_data['revol_util_percentage'] = loan_data['revol_util'].str.rstrip('%')
del loan_data['revol_util']
del loan_data['int_rate']
loan_data.head(1)

loan_amnt  funded_amnt  funded_amnt_inv  installment grade sub_grade  \
0       5000         5000         4,975.00       162.87     B        B2   

  emp_title emp_length home_ownership  annual_inc verification_status issue_d  \
0       NaN  10+ years           RENT   24,000.00            Verified  Dec-11   

  loan_status      purpose     title zip_code addr_state   dti  delinq_2yrs  \
0  Fully Paid  credit_card  Computer      860         AZ 27.65            0   

  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0           Jan-85               1         3        0      13648          9   

   out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  \
0       0.00           0.00     5,863.16         5,833.84         5,000.00   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                0.00        0.00                     0.00   

  last_pymnt_d  last_pymnt_amnt last_credit_pull_d  \
0       Jan-15           171.62             May-16   

   collections_12_mths_ex_med  pub_rec_bankruptcies term_months  \
0                        0.00                  0.00         36    

  int_rate_percentage revol_util_percentage  
0               10.65                 83.70

In [499]:
loan_data.dtypes

loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                        object
loan_status                    object
purpose                        object
title                          object
zip_code                       object
addr_state                     object
dti                           float64
delinq_2yrs                     int64
earliest_cr_line               object
inq_last_6mths                  int64
open_acc                        int64
pub_rec                         int64
revol_bal                       int64
total_acc                       int64
out_prncp                     float64
out_prncp_in

In [502]:
loan_data['term_months'].astype('int32')
loan_data['int_rate_percentage'].astype('float64')
loan_data['revol_util_percentage'].astype('float64')
loan_data['zip_code'].astype('int32')
loan_data.head(1)

loan_amnt  funded_amnt  funded_amnt_inv  installment grade sub_grade  \
0       5000         5000         4,975.00       162.87     B        B2   

  emp_title emp_length home_ownership  annual_inc verification_status issue_d  \
0       NaN  10+ years           RENT   24,000.00            Verified  Dec-11   

  loan_status      purpose     title zip_code addr_state   dti  delinq_2yrs  \
0  Fully Paid  credit_card  Computer      860         AZ 27.65            0   

  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0           Jan-85               1         3        0      13648          9   

   out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  \
0       0.00           0.00     5,863.16         5,833.84         5,000.00   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                0.00        0.00                     0.00   

  last_pymnt_d  last_pymnt_amnt last_credit_pull_d  \
0       Jan-15           171.62             May-16   

   collections_12_mths_ex_med  pub_rec_bankruptcies term_months  \
0                        0.00                  0.00         36    

  int_rate_percentage revol_util_percentage  
0               10.65                 83.70

In [507]:
loan_data['emp_experience'] = loan_data['emp_length'].str.rstrip('years')
loan_data['emp_experience'].replace({'10+': '10'}, inplace=True)
loan_data['emp_experience'][loan_data['emp_experience'] == '< 1'] = '0.5'

C:\Users\sabar\AppData\Local\Temp\ipykernel_1072\1349996822.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  loan_data['emp_experience'][loan_data['emp_experience'] == '< 1'] = '0.5'


In [508]:
del loan_data['emp_length']

In [509]:
loan_data.head(1)

loan_amnt  funded_amnt  funded_amnt_inv  installment grade sub_grade  \
0       5000         5000         4,975.00       162.87     B        B2   

  emp_title home_ownership  annual_inc verification_status issue_d  \
0       NaN           RENT   24,000.00            Verified  Dec-11   

  loan_status      purpose     title zip_code addr_state   dti  delinq_2yrs  \
0  Fully Paid  credit_card  Computer      860         AZ 27.65            0   

  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0           Jan-85               1         3        0      13648          9   

   out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  \
0       0.00           0.00     5,863.16         5,833.84         5,000.00   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         863.16                0.00        0.00                     0.00   

  last_pymnt_d  last_pymnt_amnt last_credit_pull_d  \
0       Jan-15           171.62             May-16   

   collections_12_mths_ex_med  pub_rec_bankruptcies term_months  \
0                        0.00                  0.00         36    

  int_rate_percentage revol_util_percentage emp_experience  
0               10.65                 83.70           10+